File: java/quantile-estimator/src/main/java/org/mydotey/quantileestimator/gk/GkQuantileEstimatorConfig.java
package org.mydotey.quantileestimator.gk;

import java.util.Comparator;
import java.util.Objects;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
public class GkQuantileEstimatorConfig<T> {

    private Comparator<T> _comparator;
    private double _error;
    private int _compactThreshold;

    public GkQuantileEstimatorConfig(Comparator<T> comparator, double error, int compactThreshold) {
        Objects.requireNonNull(comparator, "comparator is null");

        String format = "%s %f invalid: expected number between 0.0 and 1.0.";
        if (error < 0.0 || error > 1.0)
            throw new IllegalArgumentException(String.format(format, "error", error));

        if (compactThreshold <= 0)
            throw new IllegalArgumentException("compactThreshold is less than 1");

        _comparator = comparator;
        _error = error;
        _compactThreshold = compactThreshold;
    }

    public Comparator<T> getComparator() {
        return _comparator;
    }

    public double getError() {
        return _error;
    }

    public int getCompactThreshold() {
        return _compactThreshold;
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Gk/GkQuantileEstimatorConfig.cs
using System;
using System.Collections.Generic;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
namespace MyDotey.Quantile.Gk
{
    public class GkQuantileEstimatorConfig<T>
    {
        public virtual IComparer<T> Comparer { get; }
        public virtual double Error { get; }
        public virtual int CompactThreshold { get; }

        public GkQuantileEstimatorConfig(IComparer<T> comparer, double error, int compactThreshold)
        {
            if (comparer == null)
                throw new ArgumentNullException("comparer is null");

            String format = "{0} {1} invalid: expected number between 0.0 and 1.0.";
            if (error < 0.0 || error > 1.0)
                throw new ArgumentException(String.Format(format, "error", error));

            if (compactThreshold <= 0)
                throw new ArgumentException("compactThreshold is less than 1");

            Comparer = comparer;
            Error = error;
            CompactThreshold = compactThreshold;
        }
    }
}